### ISSUE TITLE
CA1859 doesn't consider `foreach`

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn-analyzers#6845]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1859): `Use concrete types when possible for improved performance`

### Analyzer source

Version: https://github.com/dotnet/roslyn-analyzers/commit/2a2b5f6c0d72f7ad53ebacb13bab68b05fb1e393

### Describe the bug

CA1859 doesn't warn when the field in question is `foreach`'d.

### Steps To Reproduce

```C#
public class Tests
{
    private IList<int> _list = new List<int>() { 1, 2, 3 };

    public int Sum()
    {
        int sum = 0;
        foreach (int i in _list) sum += i;
        return sum;
    }
}
```

### Expected behavior

CA1859 should issue a diagnostic that `_list`'s type could be changed to be `List<T>`.

### Actual behavior

No diagnostic.

**Comment by geeknoid** [comment]
@mavasani 

I added some code in [Update CA1859 to understand foreach statements · geeknoid/roslyn-analyzers@da0f924 (github.com)](https://github.com/geeknoid/roslyn-analyzers/commit/da0f9243db2a4c20d44b7cba692d35e53fbc915b) but I'm missing a bit of logic. I need to get the GetEnumerator method symbol that will be invoked by the generated foreach code. I can see in the debugger that the ForEachOperation type includes an Info property that is of type ForEachStatementInfo which contains the IMethodSymbol for the GetEnumerator method. Alas, the Info property is not public. The code I have now doesn't work, and manually discovering the GetEnumerator method is pretty involved since the foreach resolution logic is pretty complex.

 

I see some related stuff in the CharpDoNotCopyValue type in the CheckForEachGetEnumerator method. But that seems to be C#-specific logic which appears not to work from the assembly where CA1859 is implemented.

 

Any recommendation on how to proceed would be appreciated.
